src/routes/maven.rs

```diff
@@ -0,0 +1,262 @@
+                        .map_or("", |x| &x.version_number)
+                        .to_string(),
+                    versions: Versions {
+                        versions: version_names
+                            .iter()
+                            .map(|x| x.version_number.clone())
+                            .collect::<Vec<_>>(),
+                    },
+                    last_updated: data.inner.updated.format("%Y%m%d%H%M%S").to_string(),
+                },
+            };
+
+            // return Ok(HttpResponse::Ok().content_type("text/xml").body(response));
+            Ok(HttpResponse::Ok()
+                .content_type("text/xml")
+                .body(yaserde::ser::to_string(&respdata).map_err(|e| ApiError::XmlError(e))?))
+        } else {
+            Ok(HttpResponse::NotFound().body(""))
+        }
+    } else {
+        Ok(HttpResponse::NotFound().body(""))
+    }
+}
+
+#[get("maven/modrinth/{id}/{versionnum}/{file}")]
+pub async fn version_file(
+    req: HttpRequest,
+    web::Path((string, vnum, file)): web::Path<(String, String, String)>,
+    pool: web::Data<PgPool>,
+) -> Result<HttpResponse, ApiError> {
+    let id_option: Option<ModId> = serde_json::from_str(&*format!("\"{}\"", string)).ok();
+
+    let mut mod_data;
+
+    if let Some(id) = id_option {
+        mod_data = database::models::Mod::get_full(id.into(), &**pool)
+            .await
+            .map_err(|e| ApiError::DatabaseError(e.into()))?;
+
+        if mod_data.is_none() {
+            mod_data = database::models::Mod::get_full_from_slug(&string, &**pool)
+                .await
+                .map_err(|e| ApiError::DatabaseError(e.into()))?;
+        }
+    } else {
+        mod_data = database::models::Mod::get_full_from_slug(&string, &**pool)
+            .await
+            .map_err(|e| ApiError::DatabaseError(e.into()))?;
+    }
+
+    let user_option = get_user_from_headers(req.headers(), &**pool).await.ok();
+
+    if let Some(data) = mod_data {
+        let mut authorized = !data.status.is_hidden();
+
+        if let Some(user) = user_option {
+            if !authorized {
+                if user.role.is_mod() {
+                    authorized = true;
+                } else {
+                    let user_id: database::models::ids::UserId = user.id.into();
+
+                    let mod_exists = sqlx::query!(
+                        "SELECT EXISTS(SELECT 1 FROM team_members WHERE team_id = $1 AND user_id = $2)",
+                        data.inner.team_id as database::models::ids::TeamId,
+                        user_id as database::models::ids::UserId,
+                    )
+                    .fetch_one(&**pool)
+                    .await
+                    .map_err(|e| ApiError::DatabaseError(e.into()))?
+                    .exists;
+
+                    authorized = mod_exists.unwrap_or(false);
+                }
+            }
+        }
+
+        if authorized {
+            let vid_option = sqlx::query!(
```

Comment:
One thing that is more dependant on code rot, but when I originally put out the file format, things like requests to the database were meant to be on the models directory with the actual model struct.

Just know that I may change that in the future, and that will make the code far more readable.